proxy/envoy/watcher.go

```diff
@@ -97,13 +103,52 @@ func (w *watcher) Reload() {
 	w.agent.ScheduleConfigUpdate(config)
 }
 
-// watchCerts watches a certificate directory and calls the provided
+type watchFileEventsFn func(ctx context.Context, wch <-chan *fsnotify.FileEvent,
+	maxDelay time.Duration, notifyFn func())
+
+// watchFileEvents watches for changes on a channel and notifies via notifyFn().
+// The function batches changes so that related changes are processed together.
+// The function ensures that notifyFn() is called no more than one time per maxDelay.
+// The function does not return until the the context is cancelled.
+func watchFileEvents(ctx context.Context, wch <-chan *fsnotify.FileEvent, maxDelay time.Duration, notifyFn func()) {
```

Comment:
How about using the following code? Ticker is part of the go library, which buffers only one timer event and drops the rest

	var ev *fsnotify.FileEvent
	// ticker and channel for managing maxDelay.
	var timeChan <-chan time.Time
	ticker := time.NewTicker(maxDelay)
	defer ticker.Stop()

	for {
		select {
		case ev = <-wch:
			glog.Infof("watchFileEvents: %s", ev.String())
			timeChan = ticker.C
		case <-timeChan:
			// wait for next event
			timeChan = nil

			glog.Info("watchFileEvents: notifying")
			notifyFn()
		case <-ctx.Done():
			glog.V(2).Info("watchFileEvents has terminated")
			return
		}
	}

Reply:
great suggestion. I have updated the code.

Reply:
The approach seems fine. It doesn't eliminate the race fully. You can receive an FS event at 5s-\epsilon and capture inconsistent file system state at the next ticker. But the chances are pretty slim.